# 4.4. TASK
# Table of Contents
## 1. Import libraries and data sets
## 2. Data Wrangling Procedures
## 3. Task directions
### 3.1. Change df_ords identifier variable format
### 3.2. Change variable name
### 3.3. Frequency of order hours
### 3.4. Create data dictionary and find matching value for department_id
### 3.5. Subset for breakfast products
### 3.6. Subset for products in the alcohol, deli, beverages and meat/seafood departments
### 3.7. Number of rows of new df
### 3.8. Information about user_id 1
### 3.9. User_id 1 stats
### 3.10. Export dataframes

---

## 1. Import libraries and data sets

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
# Create path
path = r'/Users/mariliamachado/Documents/CareerFoundry/11-2024 Instacart Basket Analysis'

In [6]:
# Import Orders data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [8]:
# Import Products data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## 2. Data Wrangling Procedures

In [30]:
# Drop eval_set column
df_ords = df_ords.drop(columns = ['eval_set'])


In [14]:
# Check missing values of the days_since_prior_order column
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [16]:
# Rename order_dow column
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [32]:
# Check if column name was changed
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [20]:
# Change data type of order_id
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [22]:
# Check if data type was changed
df_ords['order_id'].dtype

dtype('O')

In [26]:
# Import departments data set
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [28]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [34]:
# Transposing departments data set
df_dep_t = df_dep.T

In [36]:
# Check if transposing worked
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [40]:
# Add index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [42]:
new_header = df_dep_t.iloc[0]

In [46]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [48]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [52]:
# set the header row as the df header
df_dep_t_new.columns = new_header

In [54]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [56]:
# Create data dictionary for departments
data_dict = df_dep_t_new.to_dict('index')

In [58]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [60]:
print(data_dict.get('19'))


{'department': 'snacks'}


In [62]:
# Subset of snacks department in df_prods
df_snacks =  df_prods[df_prods['department_id']==19]


In [64]:
df_prods['department_id']==19


0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [66]:
# save df_snacks as dataframe
df_snacks =  df_prods[df_prods['department_id']==19]


In [68]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [70]:
# second method
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]


In [72]:
# third method
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]


## 3. Task directions

### 3.1. Change df_ords identifier variable format

2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [115]:
# Check data set columns data types
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [117]:
# Change user_id data type to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [119]:
# Confirm the data type was changed
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

### 3.2. Change variable name

3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [121]:
df_ords.rename(columns = {'order_hour_of_day' : 'order_time_hour'}, inplace = True)


### 3.3. Frequency of order hours

4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [130]:
df_ords['order_time_hour'].value_counts(dropna = True)

order_time_hour
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour is 10 am.

### 3.4. Create data dictionary and find matching value for department_id

5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.


In [141]:
print(data_dict.get('4'))

{'department': 'produce'}


I had already created the data dictionary previously, so I just checked what the value 4 referred to and the answer is produce.

### 3.5. Subset for breakfast products

6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [145]:
#Breakfast subset (I checked the id for breakfast in previous steps of data wrangling)
df_breakfast =  df_prods[df_prods['department_id']==14]

In [148]:
# Check if the subset is correct
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### 3.6. Subset for products in the alcohol, deli, beverages and meat/seafood departments

7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [150]:
# Create subset for dinner parties products (once again I checked the ids in previous steps)
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [152]:
# Check subset
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### 3.7. Number of rows of new df

8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [155]:
df_dinner_parties.shape

(7650, 5)

The last dataframe has 7650 rows.

### 3.8. Information about user_id 1

9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [186]:
df_user_1 = df_ords[df_ords['user_id'] == '1']

In [188]:
df_user_1

order_id user_id  order_number  orders_day_of_week  order_time_hour  \
0   2539329       1             1                   2                8   
1   2398795       1             2                   3                7   
2    473747       1             3                   3               12   
3   2254736       1             4                   4                7   
4    431534       1             5                   4               15   
5   3367565       1             6                   2                7   
6    550135       1             7                   1                9   
7   3108588       1             8                   1               14   
8   2295261       1             9                   1               16   
9   2550362       1            10                   4                8   
10  1187899       1            11                   4                8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 3.9. User_id 1 stats

10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?


This user made 11 orders in total, 4 of them in the 4th day of the week (Thursday) and more than half of their orders were done in the morning.

11. Check the organization and structure of your notebook. Be sure to include section headings and code comments.

Done.

### 3.10. Export dataframes

12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [195]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [198]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))

14. Save your Jupyter notebook and submit it here for your tutor to review.
Done